File: tools/python-bench/worker.py
import json
from workers import Response, WorkerEntrypoint, fetch

class Default(WorkerEntrypoint):
    async def fetch(self, request):
        args = await request.json()
        name = args.get('name')
        if not isinstance(name, str) or not 1 <= len(name) <= 128:
            return Response('invalid name', status=422)
        if not self.env.UPSTREAM:
            return Response.from_json({'result': 'Hello, ' + name})
        response = await fetch(self.env.UPSTREAM, method='POST',
            headers={'content-type':'application/json'}, body=json.dumps({'name':name}))
        if not response.ok:
            raise ValueError('upstream failed')
        price = await response.json()
        if (price.get('customer') != name or type(price.get('unit_price_cents')) is not int
            or type(price.get('stock')) is not int or price['stock'] < 2 or not isinstance(price.get('currency'), str)):
            raise ValueError('invalid price')
        return Response.from_json({'result': {'customer':name,
            'total_cents':price['unit_price_cents'] * 2, 'currency':price['currency'], 'trace':name}})

File: tools/python-bench/worker.ts
// Same validation, JSON transformation and upstream work as worker.py.
export default {
  async fetch(request: Request, env: {UPSTREAM: string}) {
    const args = await request.json() as {name?: unknown};
    const name = args.name;
    if (typeof name !== 'string' || name.length < 1 || name.length > 128)
      return new Response('invalid name', {status:422});
    if (!env.UPSTREAM) return Response.json({result:'Hello, ' + name});
    const response = await fetch(env.UPSTREAM, {method:'POST',
      headers:{'content-type':'application/json'}, body:JSON.stringify({name})});
    if (!response.ok) throw new Error('upstream failed');
    const price = await response.json() as {customer: string; unit_price_cents: number; stock: number; currency: string};
    if (price.customer !== name || !Number.isInteger(price.unit_price_cents)
        || !Number.isInteger(price.stock) || price.stock < 2 || typeof price.currency !== 'string')
      throw new Error('invalid price');
    return Response.json({result:{customer:name, total_cents:price.unit_price_cents * 2, currency:price.currency, trace:name}});
  },
};
